### ISSUE TITLE
Enabling persisting notebook kernel information in notebooks

### ISSUE THREAD

**Issue by mjbvz** [microsoft/vscode#130602]

**Comment by mjbvz** [comment]
To test: 

1. Using an extension with proposed api enabled
2. Try setting the kernelspec of the active notebook using the new api:

```ts
		const ext = vscode.extensions.getExtension('vscode.ipynb');
		const api = await ext?.activate();

		if (vscode.window.activeNotebookEditor) {
			api.setKernelSpec(vscode.window.activeNotebookEditor.document.uri, { test: 'bla' });
		}
```

3. Save
3. Now reopen the editor with the text editor and confirm that kernelspec is set

**Comment by mjbvz** [comment]
Hm, not sure this api is doing the correct thing after following those steps again.

Running against https://github.com/microsoft/vscode/blob/1d8a91d86fe0ebd598642fdf71e4927a34a4f466/extensions/vscode-api-tests/testWorkspace/test.ipynb#L2, results in:

```
{
 "cells": [ ...
]
 "metadata": {
  "interpreter": {
   "hash": "815c6b7592bf74925ca002a1774bcf064bae9d6a27e7933fd9109275fb484258"
  },
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.9.5 64-bit ('myvenv': venv)"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4,
 "kernelspec": {
  "test": "bla"
 }
}
```

The notebook should also be marked dirty after its metadata is changed. It currently is not

@roblourens @jrieken Are you more familiar with the notebook editing api? Is it expected that `replaceNotebookMetadata` would not mark the targeted notebook as dirty?

**Comment by jrieken** [comment]
> @roblourens @jrieken Are you more familiar with the notebook editing api? Is it expected that replaceNotebookMetadata would not mark the targeted notebook as dirty?

It depends on the notebook serializer - they can mark what metadata is transient and what is persisted. This is how it is defined for our ipynb extension: https://github.com/microsoft/vscode/blob/1dae051c56961b868cfaaeb8db9d1893ebaafd6c/extensions/ipynb/src/ipynbMain.ts#L11

**Comment by mjbvz** [comment]
Fixed by #131681 and ported to the release branch 

See original verification steps: https://github.com/microsoft/vscode/issues/130602#issuecomment-905095132
